firmware/motion.cpp

```diff
@@ -11,101 +16,258 @@ void motion_control_arc();
+  T = (float)dy * YSTEPS_PER_MM / f * RATE_FACTOR;
+  rate[1] = (float)T  / dy / 2.0;
+#ifdef Z_DRIVER_ENABLED
+  T = (float)dz * ZSTEPS_PER_MM / f * RATE_FACTOR;
+  rate[2] = (float)T  / dz / 2.0;
+#endif
+
+#ifdef DEBUG_OUTPUT
+  Serial.print( F("// dx: ") ); Serial.print( dx ); Serial.print( " (" ); Serial.print( dx / XSTEPS_PER_MM ); Serial.print( "mm)" );
+  Serial.print( F(" dy: ") ); Serial.print( dy ); Serial.print( " (" ); Serial.print( dy / YSTEPS_PER_MM ); Serial.print( "mm)" );
+  Serial.print( F(" d: " ) ); Serial.print( sqrt( pow( dx, 2 ) + pow( dy, 2 ) ) ); Serial.print( " (" ); Serial.print( sqrt( pow( dx, 2 ) + pow( dy, 2 ) ) / XSTEPS_PER_MM ); Serial.print( "mm)" );
+  Serial.print( F(" ms/step: ") ); Serial.print( XSTEPS_PER_MM / f * 60 * 1000 );
+  Serial.print( F(" T: ") ); Serial.print( T ); Serial.println( F("ms") );
+#endif
+
+
+
+#ifdef DEBUG_OUTPUT
+  Serial.print( "// " );
+  Serial.print( rate[0] ); Serial.print( F(" RF*ms / X step, ") );
+  Serial.print( rate[1] ); Serial.println( F(" RF*ms / Y step") );
+  Serial.print( "// " );
+#endif
+  if ( destination[0] > position[0] ) {
+    direction[0] = 1;
+    WRITE( X_DIR_PIN, X_POS );
+  } else {
+    direction[0] = -1;
+    WRITE( X_DIR_PIN, X_NEG );
+  }
+
+  if ( destination[1] > position[1] ) {
+    direction[1] = 1;
+    WRITE( Y_DIR_PIN, Y_POS );
+  } else {
+    direction[1] = -1;
+    WRITE( Y_DIR_PIN, Y_NEG );
+  }
+#ifdef Z_DRIVER_ENABLED
+  if ( destination[2] > position[2] ) {
+    direction[2] = 1;
+    WRITE( Z_DIR_PIN, Z_POS );
+  } else {
+    direction[2] = -1;
+    WRITE( Z_DIR_PIN, Z_NEG );
+    ms_since_step[2] = 0;
+  }
+#endif
+
+  ms_since_step[0] = 0;
+  ms_since_step[1] = 0;
+  motion_state = MS_G0_1;
 }
 
 void motion_control_home_primary() {
-	uint8_t x = digitalRead( X_STOP_PIN );
-	uint8_t y = digitalRead( Y_STOP_PIN );
-
-	digitalWrite( X_DIR_PIN, HIGH );
-	digitalWrite( Y_DIR_PIN, HIGH );
-
-	if( x == X_STOP_ENGAGED && y == Y_STOP_ENGAGED ) {
-		motion_state = MS_HOMING_2;
-		return;
-	}
-
-	if( x != X_STOP_ENGAGED ) {
-		if( ms_since_step[0] > 0 ) {
-			enable_motors();
-			digitalWrite( X_STEP_PIN, !digitalRead( X_STEP_PIN ) );
-			ms_since_step[0] = 0;
-		}
-	}
-	if( y != Y_STOP_ENGAGED ) {
-		if( ms_since_step[1] > 0 ) {
-			enable_motors();
-			digitalWrite( Y_STEP_PIN, !digitalRead( Y_STEP_PIN ) );
-			ms_since_step[1] = 0;
-		}
-	}
+  uint8_t x = READ( X_STOP_PIN );
+  uint8_t y = READ( Y_STOP_PIN );
+
+  WRITE( X_DIR_PIN, X_NEG );
+  WRITE( Y_DIR_PIN, Y_NEG );
+#ifdef Z_DRIVER_ENABLED
+  uint8_t z = READ( Z_STOP_PIN );
+  WRITE( Z_DIR_PIN, Z_NEG );
+  if (  z == Z_STOP_ENGAGED &&  x == X_STOP_ENGAGED && y == Y_STOP_ENGAGED ) {
+#else
+  if (  x == X_STOP_ENGAGED && y == Y_STOP_ENGAGED ) {
+#endif
+    motion_state = MS_HOMING_2;
+    return;
+  }
+
+#ifdef Z_DRIVER_ENABLED
+  if ( z != Z_STOP_ENGAGED ) {
+    if ( ms_since_step[2] > 0 ) {
+      enable_motors();
+      WRITE( Z_STEP_PIN, !READ( Z_STEP_PIN ) );
```

Comment:
I do not know of implied support for toggle in arduino. I avoided to use flag variables in order to save memory. Either we use memory to keep track of the toggle status or we waste a read.
Now, the time is not  that critical cause it is taken in consideration as part of the time to the next half step.